lib/src/main/scala/spinal/lib/com/jtag/sim/JtagVpi.scala

```diff
@@ -0,0 +1,158 @@
+package solsoc
+
+import spinal.core.sim._
+import spinal.lib.com.jtag.Jtag
+
+import java.nio.{ByteBuffer, ByteOrder}
+import spinal.lib.com.jtag.sim.JtagDriver
+
+import java.io.{DataInputStream, DataOutputStream}
+import java.net.ServerSocket
+
+/*
+    * This is a JTAG VPI interface for openocd
+    * It is used to connect openocd to a JTAG interface over TCP
+    * the reference interface implementation is in https://github.com/openocd-org/openocd/blob/master/src/jtag/drivers/jtag_vpi.c
+    * Example openocd config:
+
+source [find interface/jtag_vpi.cfg]
+debug_level 1
+set _CHIPNAME SOC
+jtag newtap $_CHIPNAME cpu -irlen 5 -expected-id 0x10002fff
+# init the CPU
+target create $_CHIPNAME.cpu riscv -chain-position $_CHIPNAME.cpu
+init
+ */
+
+object JtagVpi {
+  def apply(jtag: Jtag, port: Int = 5555, jtagClkPeriod: Long) = fork {
+    val driver = JtagDriver(jtag, jtagClkPeriod)
+    var inputStream: DataInputStream = null
+    var outputStream: DataOutputStream = null
+    class SocketThread extends Thread {
+      val socket = new ServerSocket(port)
+
+      override def run(): Unit = {
+        println("VPI Server started, waiting for connection...")
+        while (true) {
+          val connection = try {
+            socket.accept()
+          } catch {
+            case e: Exception => return
+          }
+          //connection.setTcpNoDelay(true)
+          outputStream = new DataOutputStream(connection.getOutputStream())
+          inputStream = new DataInputStream(connection.getInputStream())
+          println("VPI Client connected")
+        }
+      }
+    }
+    val server = new SocketThread
+    onSimEnd(server.socket.close())
+    server.start()
+
+    // read the first message from the client and print it
+    val buffer = new Array[Byte](MaxSizeOfVpiCmd)
+    while (true) {
+      if(inputStream == null)
+      {
+        sleep(jtagClkPeriod)
+      }
+      else {
+        inputStream.readFully(buffer)
```

Comment:
I first thought that this would not be threadsafe, but after reading some of the Java docs it looks to me like there is no issue here since all of the read/write operating are done on the current thread. (i.e. that under the hood `readFully` and the `write` below directly call the appropriate functions of the socket under the hood)?